agent/src/com/cloud/agent/Agent.java

```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
             }
         }
     }
+
+    public class HostLBCheckerTask extends ManagedContextTimerTask {
+
+         @Override
+         protected void runInContext() {
+             try {
+                 int index = 0;
+                 final String[] msList = _shell.getHosts();
+                 String preferredHost  = msList[index];
+                 final String connectedHost = _shell.getConnectedHost();
+                 if (s_logger.isTraceEnabled()) {
+                     s_logger.trace("Running host LB checker task, connected host=" + connectedHost + ", preferred host=" + preferredHost);
+                 }
+                 final long wait = _shell.getFailbackPollingWait();
+                 while (!preferredHost.equals(connectedHost)) {
+                     if (_link != null) {
+                         boolean isHostUp = true;
+                         try (final Socket socket = new Socket()) {
+                             socket.connect(new InetSocketAddress(preferredHost, _shell.getPort()), 5000);
+                         } catch (final IOException e) {
+                             isHostUp = false;
+                             if (s_logger.isDebugEnabled()) {
+                                 s_logger.debug("Host: " + preferredHost + " is not reachable, trying to reconnect to next host...");
+                             }
+                         }
+                         if (isHostUp && _link != null) {
+                             long initialTime = System.currentTimeMillis();
+                             while (System.currentTimeMillis() - initialTime < wait && _inProgress.get() > 0) {
+                                 s_logger.debug("Waiting for active commands to be completed");
+                                 Thread.sleep(5000l);
+                             }
+                             if (_inProgress.get() == 0) {
+                                 if (s_logger.isDebugEnabled()) {
+                                     s_logger.debug("Preferred host is found to be reachable, trying to reconnect to: " + preferredHost);
+                                 }
+                                 _shell.resetHostCounter();
+                                 reconnect(_link);
```

Comment:
@nvazquez just so we're on the same page, when performing a reconnect after resting the host counter, `reconnect` will start getting host from the first in the list, i.e. the preferred one.

Reply:
Right, my bad. Now it makes sense as timer task will attempt reconnection to the first host in the list